python/revng/scripting/projects/daemon_project.py

```diff
@@ -0,0 +1,163 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import json
+from typing import List
+
+import yaml
+from gql import Client, gql
+from gql.transport.aiohttp import AIOHTTPTransport
+
+from revng.model import Binary, Function, dump_diff_to_string, load_model_from_string, make_diff  # type: ignore[attr-defined] # noqa: E501
+
+from .project import Project
+
+
+class DaemonProject(Project):
+    def __init__(self, url: str):
+        self.transport: AIOHTTPTransport = AIOHTTPTransport(url=f"{url}/graphql/")
+        self.client: Client = Client(
+            transport=self.transport, fetch_schema_from_transport=True, execute_timeout=None
+        )
+        super().__init__()
+
+    def import_binary(self, input_binary_path: str) -> None:
+        query = gql(
+            """
+            mutation upload($file: Upload!) {
+                uploadFile(file: $file, container: "input")
+            }
+        """
+        )
+
+        with open(input_binary_path, "rb") as binary_file:
+            result = self.client.execute(
+                query, variable_values={"file": binary_file}, upload_files=True
+            )
+
+        if not result["uploadFile"]:
+            raise RuntimeError("File upload failed")
+
+    def _get_artifact(self, artifact_name: str, targets_class: List[Function] = []) -> bytes | str:
+        if not targets_class:
+            targets_class = self.model.Functions
+        targets = [tc.key() for tc in targets_class]
+
+        result = self._get_artifact_internal(artifact_name, targets)
+        return self._get_result_mime(artifact_name, result)
+
+    def _get_artifact_internal(
+        self, artifact_name: str, targets_class: List[str] = []
+    ) -> bytes | str:
+        query = gql(
+            """
+            query($step: String!, $paths: String!, $index: BigInt!) {
+                produceArtifacts(step: $step, paths: $paths, index: $index) {
+                    ... on Produced {
+                        result
+                    }
+                }
+            }
+        """
+        )
+        variables = {
+            "step": artifact_name,
+            "paths": ",".join(targets_class),
+            "index": self._get_content_commit_index(),
+        }
+
+        response = self.client.execute(query, variable_values=variables)
+        return response["produceArtifacts"]["result"].encode()
+
+    def analyze(self, analysis_name: str, targets=dict[str, List[str]], options: dict = {}) -> None:
```

Comment:
There's still `targets=Dict[str, List[str]]`, I think you meant `targets: Dict[str, List[str]] = {}`?

Reply:
Now it's as I think it should be. The default arg is `Dict[str, List[str]]` and then in the `cli` we pass and empty dict because it's not implemented yet.

Reply:
I'm a bit confused, AFAIK `Dict[str, List[str]]` is a **type** notation and it should not be used as a value.

Reply:
Sorry I made a typo, I meat `type` instead of `arg`. I meant to say that in the abstract method we expect the `targets` to not be empty and the in the `cli` we pass an empty `targets` since they are not supported yet by the `cli`.

Reply:
Ah ok, correct to `targets: Dict[str, List[str]]` and the resolve this.